arch/lkl/kernel/syscalls.c

```diff
@@ -33,400 +32,136 @@ syscall_handler_t syscall_table[__NR_syscalls] = {
-		if (!data) {
-			data = __lkl_create_syscall_thread();
-			if (!data)
-				lkl_puts("failed to create syscall thread\n");
+		if (!task) {
+			ret = new_host_task(&task);
+			if (ret)
+				goto out;
 		}
-	}
-	if (!data || no == __NR_reboot)
-		data = &default_syscall_thread_data;
-
-	return __lkl_syscall(data, no, params);
-}
-
-static asmlinkage long
-sys_create_syscall_thread(struct syscall_thread_data *data)
-{
-	pid_t pid;
-
-	pid = kernel_thread(syscall_thread, data, CLONE_VM | CLONE_FS |
-			CLONE_FILES | CLONE_THREAD | CLONE_SIGHAND | SIGCHLD);
-	if (pid < 0)
-		return pid;
-
-	return 0;
-}
-
-
-/*
- * A synchronization algorithm between cleanup_syscall_threads (which terminates
- * all remaining syscall threads) and destructors functions (which frees a
- * syscall thread as soon as the associated host thread terminates) is required
- * since destructor functions run in host context and is not subject to kernel
- * scheduling.
- *
- * An atomic counter is used to count the number of running destructor functions
- * and allows the cleanup function to wait for the running destructor functions
- * to complete.
- *
- * The cleanup functions adds MAX_SYSCALL_THREADS to this counter and this
- * allows the destructor functions to check if the cleanup process has started
- * and abort execution. This prevents "late" destructors from trying to free the
- * syscall threads.
- *
- * This algorithm assumes that we never have more the MAX_SYSCALL_THREADS
- * running.
- */
-#define MAX_SYSCALL_THREADS 1000000
-static unsigned int destrs;
-
-/*
- * This is called when the host thread terminates if auto_syscall_threads is
- * enabled. We use it to remove the associated kernel syscall thread since it is
- * not going to be used anymore.
- *
- * Note that this run in host context, not kernel context.
- *
- * To avoid races between the destructor and lkl_sys_halt we announce that a
- * destructor is running and also check to see if lkl_sys_halt is running, in
- * which case we bail out - the kernel thread is going to be / has been stopped
- * by lkl_sys_halt.
- */
-static void syscall_thread_destructor(void *_data)
-{
-	struct syscall_thread_data *data = _data;
 
-	if (!data)
-		return;
+		lkl_ops->tls_set(task_key, task);
 
-	if (__sync_fetch_and_add(&destrs, 1) < MAX_SYSCALL_THREADS)
-		__lkl_stop_syscall_thread(data, true);
-	__sync_fetch_and_sub(&destrs, 1);
-}
+		switch_to_host_task(task);
+	}
 
-static void cleanup_syscall_threads(void)
-{
-	struct syscall_thread_data *i = NULL, *aux;
+	ret = run_syscall(no, params);
 
-	/* announce destructors that we are stopping */
-	__sync_fetch_and_add(&destrs, MAX_SYSCALL_THREADS);
+	if (count > 1) {
+		set_current_state(TASK_UNINTERRUPTIBLE);
+		if (!thread_set_sched_jmp())
+			schedule();
+		count--;
+		return ret;
+	}
 
-	/* wait for any pending destructors to complete */
-	while (__sync_fetch_and_add(&destrs, 0) > MAX_SYSCALL_THREADS)
-		schedule_timeout(1);
+out:
```

Comment:
Yes, you are right. I am changing this in next version of the path, to call schedule from lkl_trigger_irq() if need_resched() returns true.